snowfall/training/compute_expected_times.py

```diff
@@ -0,0 +1,177 @@
+                                      treat_epsilons_specially=False)
+
+    paths_lats = k2.intersect_dense(path_decoding_graphs,
+                                    dense_fsa_vec,
+                                    output_beam=10.0,
+                                    a_to_b_map=path_to_seq_map,
+                                    seqframe_idx_name='seqframe_idx')
+
+    # by seq we mean the original sequence indexes, by path we mean the indexes
+    # of the n-best paths; path_to_seq_map maps from path-index to seq-index.
+    seqs_shape = dense_fsa_vec.dense_fsa_vec.shape()
+
+    # paths_shape will also be a k2.RaggedInt with 2 axes
+    paths_shape, _ = k2.ragged.index(seqs_shape,
+                                     path_to_seq_map,
+                                     need_value_indexes=False)
+
+    seq_starts = seqs_shape.row_splits(1)[:-1]
+    path_starts = paths_shape.row_splits(1)[:-1]
+
+    # We can map from seqframe_idx  for paths, to seqframe_idx for seqs,
+    # by adding path_offsets.  path_offsets is indexed by path-index.
+    path_offsets = path_starts - k2.index(seq_starts, path_to_seq_map)
+
+    # assign new attribute 'pathframe_idx' that combines path and frame.
+    paths_lats_arc2path = k2.index(paths_lats.arcs.shape().row_ids(1),
+                                   paths_lats.arcs.shape().row_ids(2))
+
+    paths_lats.pathframe_idx = paths_lats.seqframe_idx + k2.index(
+        path_offsets, paths_lats_arc2path)
+
+    pathframe_to_pathphone = k2.create_sparse(rows=paths_lats.pathframe_idx,
+                                              cols=paths_lats.pathphone_idx,
+                                              values=paths_lats.get_arc_post(
+                                                  True, True).exp(),
+                                              min_col_index=0)
+
+    # TODO(fangjun): this check is for test, will remove it
+    sum_per_row = torch.sparse.sum(pathframe_to_pathphone, dim=1).to_dense()
+    expected_sum_per_row = torch.ones_like(sum_per_row)
+    assert torch.allclose(sum_per_row, expected_sum_per_row)
+
+    frame_idx = torch.arange(paths_shape.num_elements()) - k2.index(
+        path_starts, paths_shape.row_ids(1))
+
+    # TODO(fangjun): we can swap `rows` and `cols`
+    # while creating `pathframe_to_pathphone` so that
+    # `t()` can be omitted here.
+    weighted_occupation = torch.sparse.mm(
+        pathframe_to_pathphone.t(),
+        frame_idx.unsqueeze(-1).to(pathframe_to_pathphone.dtype))
+
+    # sum over columns to get the total occupation
+    # Use `to_dense()` here because PyTorch does not
+    # support div(dense_matrix, sparse_matrix)
+    total_occupation = torch.sparse.sum(pathframe_to_pathphone,
+                                        dim=0).to_dense()
+
+    # TODO(fangjun): remove print
+    print('total_occupation[:50]\n', total_occupation[:50])
+
+    expected_times = weighted_occupation.squeeze() / total_occupation
```

Comment:
be aware that division by zero is a possibility here, for epsilons.

Reply:
thanks, will fix it by adding an EPS to the denominator.